README.md
# BTyper3

*In silico* taxonomic classification of *Bacillus cereus* group isolates using assembled genomes

## Overview

BTyper3 is a command-line tool for taxonomically classifying *Bacillus cereus* group genomes using a standardized nomenclature.

The program, as well as the associated databases, can be downloaded from https://github.com/lmc297/BTyper3.

Post issues at https://github.com/lmc297/BTyper3/issues.

For more information, check out the BTyper3 wiki at https://github.com/lmc297/BTyper3/wiki.

------------------------------------------------------------------------

## Installation

For more information, check out the <a href="https://github.com/lmc297/BTyper3/wiki">BTyper3 wiki</a>

### conda (recommended)

To create a conda environment named `btyper3` and install BTyper3 and all of its dependencies:

1. Install [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html), if necessary
2. Create a new environment named `btyper3` by running the following command from your terminal:
   ```console
   conda create -n btyper3
   ```
3. Activate your `btyper3` environment by running the following command from your terminal:
   ```console
   conda activate btyper3
   ```
4. Install [BTyper3](https://anaconda.org/bioconda/btyper3) by running the following command from your terminal:
   ```console
   conda install -c bioconda btyper3
   ```
5. You can now run `btyper3`! Run the following command from your terminal to view all `btyper3` options, or <a href="https://github.com/lmc297/BTyper3/wiki">check the BTyper3 wiki</a> for details:
   ```console
   btyper3 --help
   ```
6. When you're done with BTyper3, you can deactivate the `btyper3` environment by running the following:
   ```console
   conda deactivate
   ```

### pip

1. To run BTyper3, please download and install the following dependencies, if necessary:

   - [Python 3](https://www.python.org/downloads/)
   - [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download) v2.9.0+ and up
  <!-- - [Biopython](https://biopython.org/wiki/Download) v1.7.4 and up (for Python 3)
  - [Pandas](https://pandas.pydata.org/pandas-docs/stable/install.html) (for Python 3)
  - [NumPy] (for Python 3)
  - [PyFastANI] version 0.3 and up</a> -->

2. [Add BLAST+ to your path](https://unix.stackexchange.com/questions/26047/how-to-correctly-add-a-path-to-path), if necessary (to check if BLAST+ is in your path, try running `makeblastdb -h` and `tblastn -h` from your command line; you should get a help message for each command, with no error messages)

3. Install via `pip` (this will download required Python dependencies as well):
   ```console
   pip install btyper3  
   ```

------------------------------------------------------------------------

## News: updates in BTyper3 v3.2.0, v3.3.0, and v3.4.0 -- new species just dropped!

The primary function of BTyper3 is to allow users to taxonomically classify *B. cereus* group genomes using a standardized nomenclature (see <a href="https://journals.asm.org/doi/10.1128/mBio.00034-20">here</a> and <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">here</a> for details regarding how the standardized nomenclature was constructed, and how it compares to historical typing methods, respectively). However, we understand that some users may also want to compare their *B. cereus* group genomes to the type strain genomes of published *B. cereus* group species. Thus, in BTyper3 v3.2.0, we have added the `--ani_typestrains` option, which calculates ANI values between a query genome and the genomes of all published *B. cereus* group species type strains and reports the type strain that produces the highest ANI value.

The type strain genomes used by BTyper3's `--ani_typestrains` option correspond to the species discussed in <a href="https://www.tandfonline.com/doi/full/10.1080/10408398.2021.1916735">Figure 2 of our taxonomy review</a>, plus species published after the review was published</a> (i.e., *Bacillus sanguinis*, *Bacillus paramobilis*, and *Bacillus hominis*, added in v3.2.0; "*B. arachidis*" and *B. rhizoplanae*, added in v3.3.0; "*B. pretiosus*", added in v3.4.0). Within the standardized taxonomy that BTyper3 uses for genomospecies assignment:

* All members of *Bacillus sanguinis* (type strain RefSeq Assembly Accession GCF_018332475.1) belong to *B. mosaicus* (i.e., *B. sanguinis* is not considered a novel species in the standardized taxonomy)

* All members of *Bacillus paramobilis* (type strain RefSeq Assembly Accession GCF_018332495.1) belong to *B. mosaicus* (i.e., *B. paramobilis* is not considered a novel species in the standardized taxonomy)

* All members of *Bacillus hominis* (type strain RefSeq Assembly Accession GCF_018332515.1) belong to *B. mycoides* (i.e., *B. hominis* is not considered a novel species in the standardized taxonomy)

* "*Bacillus arachidis*" (type strain RefSeq Assembly Accession GCF_017498775.1) replaces the putative genomospecies previously referred to as "Unknown Species 17" in the standardized taxonomy

* *Bacillus rhizoplanae* (type strain RefSeq Assembly Accession GCF_917563915.1) represents a novel genomospecies within the standardized taxonomy and has been added to the database

* All members of "*Bacillus pretiosus*" (type strain RefSeq Assembly Accession GCF_025916425.1) belong to *B. mosaicus* (i.e., "*B. pretiosus*" is not considered a novel species in the standardized taxonomy)

**Importantly, *B. cereus* group species are often proposed in the literature using unstandardized approaches** (e.g., varying genomospecies thresholds, which may produce overlapping genomospecies). We have added the type strain comparison method in BTyper3 v3.2.0, as users may still want to compare a query genome with the type strains of published *B. cereus* group species. However, interpret results with caution, as some *B. cereus* group genomes may belong to multiple species using type strain genomes.

For more information, check out our:

* <a href="https://www.tandfonline.com/doi/full/10.1080/10408398.2021.1916735">Review of *B. cereus* group taxonomy/nomenclature</a>

* <a href="https://journals.asm.org/doi/full/10.1128/mBio.00034-20">Standardized nomenclature for the *B. cereus* group</a>

* <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">Comparison of our standardized nomenclature to other *B. cereus* group typing methods (e.g., MLST, *panC*, ANI-based comparisons to species type strain genomes)</a>

------------------------------------------------------------------------

## Citation

### If you found the BTyper3 tool, its source code, and/or any of its associated databases useful, please cite:

Carroll, Laura M., Martin Wiedmann, Jasna Kovac. 2020. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7042689/">"Proposal of a Taxonomic Nomenclature for the *Bacillus cereus* Group Which Reconciles Genomic Definitions of Bacterial Species with Clinical and Industrial Phenotypes."</a> *mBio* 11(1): e00034-20; DOI: 10.1128/mBio.00034-20.

Carroll, Laura M., Rachel A. Cheng, Jasna Kovac. 2020. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7536271/">"No Assembly Required: Using BTyper3 to Assess the Congruency of a Proposed Taxonomic Framework for the *Bacillus cereus* group with Historical Typing Methods."</a> *Frontiers in Microbiology* 11: 580691; DOI: 10.3389/fmicb.2020.580691.

------------------------------------------------------------------------


## Quick Start

For detailed information, check out the <a href="https://github.com/lmc297/BTyper3/wiki">BTyper3 wiki</a>

### Command Structure

```
btyper3 -i [fasta] -o [output directory] [options...]
```

For help, type `btyper3 -h` or `btyper3 --help`

For your current version, type `btyper3 --version`

### Sample Commands

#### Perform all default analyses, using an assembled genome (complete or draft) in (multi-)FASTA format as input (assumes fastANI is in the user's path):

```
btyper3 -i /path/to/genome.fasta -o /path/to/desired/output_directory
```

#### Perform all default analyses, using an assembled genome (complete or draft) in (multi-)FASTA format as input (fastANI is not in the user's path):

```
btyper3 -i /path/to/genome.fasta -o /path/to/desired/output_directory --fastani_path /path/to/FastANI_executable/fastANI
```

#### Perform all default analyses, plus pseudo-gene flow unit assignment, using an assembled genome (complete or draft) in (multi-)FASTA format as input (assumes fastANI is in the user's path):

```
btyper3 -i /path/to/genome.fasta -o /path/to/desired/output_directory --ani_geneflow True
```

#### Perform seven-gene MLST only, using user-supplied MLST gene sequences and the latest version of the PubMLST *B. cereus s.l.* database (sequences can be in multi-FASTA format, or concatenated into a single sequence in FASTA format):

```
btyper3 -i /path/to/mlst.fasta -o /path/to/desired/output_directory --ani_species False --ani_subspecies False --ani_typestrains False --virulence False --bt False --panC False --download_mlst_latest True
```

#### Perform *panC* group assignment only, using a user-supplied *panC* gene sequence in FASTA format:

```
btyper3 -i /path/to/panC.fasta -o /path/to/desired/output_directory --ani_species False --ani_subspecies False --ani_typestrains False --virulence False --bt False --mlst False
```

#### Perform virulence factor and Bt toxin-encoding gene detection in a plasmid sequence in FASTA format:

```
btyper3 -i /path/to/plasmid.fasta -o /path/to/desired/output_directory --ani_species False --ani_subspecies False --ani_typestrains False --mlst False --panC False
```


------------------------------------------------------------------------


Disclaimer: BTyper3 is pretty neat! However, no tool is perfect, and BTyper3 cannot definitively prove whether an isolate is pathogenic or not. As always, interpret your results with caution. We are not responsible for taxonomic misclassifications, misclassifications of an isolate's pathogenic potential or industrial utility, and/or misinterpretations (biological, statistical, or otherwise) of BTyper3 results.


CHANGELOG.md
# BTyper3 CHANGELOG

All notable changes to BTyper3 will be documented in this file.
## [3.4.0] - 2023-06-01
### Added
- Added "*Bacillus pretiosus*" (NCBI RefSeq Assembly Accession GCF_025916425.1) to the `--ani_typestrains` database
### Changed
- Changed `--ani_typestrains` to `True` by default (to be compatible with BRisk USDA project)

## [3.3.4] - 2022-12-13
### Fixed
- `btyper3.ani` occasionally crashing in `--ani_geneflow` when the best hit is under threshold and not enough hits were obtained.
### Changed
- Reorganize `btyper3.ani` data files to store metadata and rewrite formatting code in `Ani.run_fastani`.

## [3.3.3] - 2022-10-17
### Fixed
- `btyper3.ani` failing to read query genome from an exhausted sequence iterator.

## [3.3.2] - 2022-08-20
### Fixed
- PyFastANI result handling code crashing on genomes too distant from the database genomes.
### Added
- Warning when running `btyper3` on a very fragmented genome.

## [3.3.1] - 2022-08-12
### Fixed
- Issue with PyFastANI migration when running the `--ani_geneflow` and `--ani_typestrains` operations.
- Wrong program name being shown when running `btyper3` in module form (`python -m btyper3`).

## [3.3.0] - 2022-08-11
### Added
- Added "*Bacillus arachidis*" (NCBI RefSeq Assembly Accession GCF_017498775.1) to the `--ani_typestrains` database
- Added *Bacillus rhizoplanae* (NCBI RefSeq Assembly Accession GCF_917563915.1) to the `--ani_typestrains` and `--ani_species` databases
### Changed
- Changed the name of "Unknown Species 17" to "*Bacillus arachidis*" in the `--ani_species` database
- Included `ndb`, `not`, `ntf`, and `nto` in the list of BLAST database files to delete

## [3.2.0] - 2021-04-12
### Added
- Added `--ani_typestrains` option to `btyper3`, which allows users to compare a query *B. cereus* group genome to the type strain genomes of all published *B. cereus* group species in the literature; this options calculates ANI values between the query genome and the genomes of all published *B. cereus* group species type strains and reports the type strain that produces the highest ANI value.
- Added `typestrains-only` database to `build_btyper3_ani_db.py`, which allows users to download the type strain genomes of all published *B. cereus* group species; this database is used with the `--ani_typestrains` in `btyper3`. Type strain genomes correspond to the <a href="https://www.tandfonline.com/doi/full/10.1080/10408398.2021.1916735">species discussed in Figure 2 of our taxonomy review</a>, plus <a href="https://pubmed.ncbi.nlm.nih.gov/34494947/">three species</a> published after the review was published.

## [3.1.2] - 2021-04-12
### Changed
- Fixed an indentation typo in `download_pubmlst_latest.py` so that `xml` could properly parse the PubMLST xml

## [3.1.1] - 2020-10-25
### Added
- Added `download_pubmlst_latest.py` script, which users can run to download latest PubMLST database for *Bacillus cereus*, independently of executing the `btyper3` program

### Changed
- Updated `--download_mlst_latest` portion of `btyper3` code to be compatible with changes to <a href="https://pubmlst.org/data/dbases.xml">PubMLST's XML</a> format; recent changes made by PubMLST to the filenames in the XML caused previous versions of BTyper3 to produce an error and not download the database

### Removed
- Removed <a href="https://mbio.asm.org/content/11/1/e00034-20">Unknown Species 14</a> from the species reference genomes used for genomospecies assignment; using recent genomes submitted to NCBI (accessed 2020-05-14), Unknown Species 14 was found to overlap with *B. luti* at a 92.5 ANI threshold (see genomes with NCBI RefSeq Assembly Accessions GCF_008764075.1, GCF_007667455.1, and GCF_007673665.1, which were assigned to both *B. luti* and Unknown Species 14 at a 92.5 ANI threshold); Unknown Species 14 was therefore merged into *B. luti* in BTyper3 v. 3.1.1

## [3.1.0] - 2020-06-28
### Added
- Added <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">*in silico* seven-gene multi-locus sequence typing (MLST)</a>, using PubMLST's scheme for *Bacillus cereus*
- Added <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">*panC* group assignment</a> using an eight-group *panC* group assignment framework
- Added <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">virulence factor detection</a> for additional virulence factors (i.e., toxin genes associated with diarrheal *Bacillus cereus* illness, and genes associated with capsule production)
- Added <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">average-nucleotide identity (ANI)-based pseudo-gene flow unit assignment method</a>

### Changed
- Changed <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2020.580691/full">default virulence factor detection thresholds to 70 and 80% amino acid identity and coverage, respectively</a>

## [3.0.2] - 2019-11-18
### Changed
- Swapped anthrax and cereulide column names in final results file (previously, the column header for anthrax genes was above cereulide synthetase genes in the final results file, and vice versa; the typo was thus fixed)
- Edited the README to convey that BLAST+ version 2.10.0 is not compatible with BTyper3

## [3.0.1] - 2019-10-05
### Changed
- Changed default percent amino acid identity for virulence gene detection and Bt toxin gene detection from 70% to 50% to be compatible with what was reported in the <a href="https://www.biorxiv.org/content/10.1101/779199v1">preprint</a>
- Fixed pandas.read_csv error which would occur when no fastANI and/or BLAST output file would be produced, or when a file was empty

btyper3/__init__.py
#!/usr/bin/env python3

import argparse
import contextlib
import datetime
import importlib.resources
import logging
import os
import sys
import shutil
import tempfile
import urllib.request
import warnings
import xml.etree.ElementTree as etree

from Bio import SeqIO

from .ani import Ani
from .blast import Blast
from .mlst import Mlst
from .print_final_results import FinalResults

__author__ = "Laura M. Carroll <lmc297@cornell.edu>"
__version__ = "3.4.0"


@contextlib.contextmanager
def _forward_warnings():
	_showwarning = warnings.showwarning
	try:
		warnings.showwarning = lambda message, category, filename, lineno, file=None, line=None: logging.warning(f"Warning: {message}")
		yield
	finally:
		warnings.showwarning = _showwarning


def run_pipeline(args):
	"""
	run btyper3
	"""

	# get path to btyper3 executable
	btyper3_path = os.path.realpath(__file__)
	btyper3_path = btyper3_path.rpartition("/")[0].strip() + "/"

	# get input and output arguments and prefix to use for output files
	infile = args.input[0]
	prefix = infile.split("/")[-1].strip()
	prefix = ".".join(prefix.split(".")[0:-1])
	output = args.output[0]

	# add a / to output directory name, if one is not already supplied
	if not output.endswith("/"):
		output = output.strip() + "/"

	# if the output directory doesn't yet exist, make it
	if not os.path.isdir(output):
		os.mkdir(output)

	# if a btyper3_final_results directory does not exist
	# in the output directory, make it
	if not os.path.isdir(output + "btyper3_final_results"):
		os.mkdir(output + "btyper3_final_results/")

	# make a logs directory for log files
	if not os.path.isdir(output + "btyper3_final_results/logs"):
		os.mkdir(output + "btyper3_final_results/logs")

	# save path to final results directory
	final_results_directory = output + "btyper3_final_results/"

	# initialize log file
	logging.basicConfig(level = logging.DEBUG, filename = final_results_directory + "logs/" + prefix + ".log", filemode = "a+", format = "%(message)s")
	logging.getLogger().addHandler(logging.StreamHandler())


	# get arguments
	# evalue is stored as a string because it gets passed to blast command as a string
	ani_species = args.ani_species
	ani_subspecies = args.ani_subspecies
	ani_geneflow = args.ani_geneflow
	ani_typestrains = args.ani_typestrains
	virulence = args.virulence
	bt = args.bt
	mlst = args.mlst
	panC = args.panC
	vdb = args.virulence_db
	vpthresh = int(args.virulence_identity)
	vqthresh = int(args.virulence_coverage)
	bpthresh = int(args.bt_identity)
	bqthresh = int(args.bt_coverage)
	overlap = float(args.bt_overlap)
	evalue = str(args.evalue)
	download_mlst_latest = args.download_mlst_latest

	# log to file
	now = datetime.datetime.now
	logging.info("Welcome to BTyper3!")
	logging.info("You are initializing this run at " + now().strftime("%Y-%m-%d %H:%M"))
	logging.info("You ran the following command: ")
	logging.info(" ".join([str(sa) for sa in sys.argv]))
	logging.info("Report bugs/concerns to Laura M. Carroll <laura.carroll@embl.de>")


	# perform species and/or supspecies assignment
	if ani_species == "True" or ani_subspecies == "True" or ani_geneflow == "True" or ani_typestrains == "True":

		with _forward_warnings():

			if ani_species == "True":

				get_species = Ani(
					taxon = "species",
					fasta = infile,
					final_results_directory = final_results_directory,
					prefix = prefix)

				logging.info("Using PyFastANI to assign " + prefix + " to a species at " + now().strftime("%Y-%m-%d %H:%M"))
				final_species = get_species.run_fastani("species", infile, final_results_directory, prefix)
				logging.info("Finished species assignment of " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			else:

				final_species = "(Species assignment not performed)"

			if ani_subspecies == "True":

				get_subspecies = Ani(
					taxon = "subspecies",
					fasta = infile,
					final_results_directory = final_results_directory,
					prefix = prefix)

				logging.info("Using FastANI to assign " + prefix + " to a subspecies (if applicable) at " + now().strftime("%Y-%m-%d %H:%M"))
				final_subspecies = get_subspecies.run_fastani("subspecies", infile, final_results_directory, prefix)
				logging.info("Finished subspecies assignment of " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			else:

				final_subspecies = "(Subspecies assignment not performed)"

			if ani_geneflow == "True":

				get_geneflow = Ani(
					taxon = "geneflow",
					fasta = infile,
					final_results_directory = final_results_directory,
					prefix = prefix)

				logging.info("Using FastANI to assign " + prefix + " to a pseudo-gene flow unit at " + now().strftime("%Y-%m-%d %H:%M"))
				final_geneflow = get_geneflow.run_fastani("geneflow", infile, final_results_directory, prefix)
				logging.info("Finished pseudo-gene flow unit assignment of " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			else:

				final_geneflow = "(Pseudo-gene flow unit assignment not performed)"

			if ani_typestrains == "True":

				get_typestrains = Ani(
					taxon = "typestrains",
					fasta = infile,
					final_results_directory = final_results_directory,
					prefix = prefix)

				logging.info("Using FastANI to compare " + prefix + " to B. cereus s.l. species type strain genomes at " + now().strftime("%Y-%m-%d %H:%M"))
				final_typestrains = get_typestrains.run_fastani("typestrains", infile, final_results_directory, prefix)
				logging.info("Finished B. cereus s.l. species type strain comparison of " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			else:

				final_typestrains = "(Type strain-based taxonomic assignment not performed)"

	else:

		final_species = "(Species assignment not performed)"
		final_subspecies = "(Subspecies assignment not performed)"
		final_geneflow = "(Pseudo-gene flow unit assignment not performed)"
		final_typestrains = "(Type strain-based taxonomic assignment not performed)"



	# perform virulence-associated biovar assignment
	if virulence == "True":

		if vdb == "aa":
			vdb_name = "btyper3_virulence_sequences.faa"
			vdb_task = "tblastn"

		elif vdb == "nuc":
			vdb_name = "btyper3_virulence_sequences.ffn"
			vdb_task = "blastn"


		with importlib.resources.path("btyper3.seq_virulence_db", vdb_name) as vdb_path:

			get_virulence = Blast(
				task = vdb_task,
				dbseqs = infile,
				fasta = vdb_path,
				final_results_directory = final_results_directory,
				prefix = prefix,
				suffix = "virulence",
				pthresh = vpthresh,
				qthresh = vqthresh,
				overlap = overlap,
				evalue = evalue)

			logging.info("Using " + vdb_task + " to identify potential virulence factors in " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			vir = get_virulence.run_blast(vdb_task, infile, vdb_path, final_results_directory, prefix, "virulence", evalue)
			anthracis, emetic, nhe, hbl, cytK, sph, cap, has, bps = get_virulence.parse_virulence(vir, vpthresh, vqthresh)

			logging.info("Finished virulence factor detection in " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))


	else:

		anthracis = "(Virulence factor detection not performed)"
		emetic = "(Virulence factor detection not performed)"
		nhe = "(Virulence factor detection not performed)"
		hbl = "(Virulence factor detection not performed)"
		cytK = "(Virulence factor detection not performed)"
		sph = "(Virulence factor detection not performed)"
		cap = "(Virulence factor detection not performed)"
		has = "(Virulence factor detection not performed)"
		bps = "(Virulence factor detection not performed)"



	# perform Thuringiensis biovar assignment
	if bt == "True":

		with importlib.resources.path("btyper3.seq_bt_db", "btyper3_bt_sequences.faa") as bt_path:

			get_bt = Blast(
				task = "tblastn",
				dbseqs = infile,
				fasta = bt_path,
				final_results_directory = final_results_directory,
				prefix = prefix,
				suffix = "bt",
				pthresh = bpthresh,
				qthresh = bqthresh,
				overlap = overlap,
				evalue = evalue)

			logging.info("Using tblastn to identify potential Bt genes in " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			bt_results = get_bt.run_blast("tblastn", infile, bt_path, final_results_directory, prefix, "bt", evalue)
			bt_final = get_bt.parse_bt(bt_results, bpthresh, bqthresh, overlap)

			logging.info("Finished Bt toxin gene detection for " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))


	else:

		bt_final = "(Bt toxin gene detection not performed)"




	# perform multi-locus sequence typing (MLST) using PubMLST's seven-gene scheme for Bacillus cereus
	if mlst == "True":

		with contextlib.ExitStack() as ctx:

			if download_mlst_latest == "True":

				logging.info("Downloading most recent PubMLST datbase at " + now().strftime("%Y-%m-%d %H:%M"))
				with urllib.request.urlopen("https://pubmlst.org/data/dbases.xml") as req:
					tree = etree.parse(req)
					parent = next(e for e in tree.iter("species") if e.text.strip() == "Bacillus cereus")
					urls = (e.text for e in parent.iter("url"))

				mlst_file = ctx.enter_context(tempfile.NamedTemporaryFile(suffix=".fas", mode="wb", buffering=0))
				bcereus_file = ctx.enter_context(tempfile.NamedTemporaryFile(suffix=".txt", mode="wb", buffering=0))
				mlst_path = mlst_file.name
				bcereus_path = bcereus_file.name

				for url in urls:
					if "alleles_fasta" in url:
						with urllib.request.urlopen(url) as req:
							shutil.copyfileobj(req, mlst_file)
					elif "profiles_csv" in url:
						with urllib.request.urlopen(url) as req:
							shutil.copyfileobj(req, bcereus_file)

				logging.info("Finished downloading most recent PubMLST datbase at " + now().strftime("%Y-%m-%d %H:%M"))

			else:

				db_time = importlib.resources.read_text("btyper3.seq_mlst_db", "timestamp.txt").strip()
				logging.info("Using local PubMLST database (downloaded at {})".format(db_time))

				mlst_path = ctx.enter_context(importlib.resources.path("btyper3.seq_mlst_db", "mlst.fas"))
				bcereus_path = ctx.enter_context(importlib.resources.path("btyper3.seq_mlst_db", "bcereus.txt"))


			get_mlst = Blast(
				task = "blastn",
				dbseqs = infile,
				fasta = mlst_path,
				final_results_directory = final_results_directory,
				prefix = prefix,
				suffix = "mlst",
				pthresh = 0,
				qthresh = 0,
				overlap = overlap,
				evalue = evalue)

			logging.info("Using blastn to identify potential seven-gene MLST genes in " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			mlst_results = get_mlst.run_blast("blastn", infile, mlst_path, final_results_directory, prefix, "mlst", evalue)
			mlst_alleles, perfect_matches = get_mlst.parse_mlst(mlst_results)

			logging.info("Finished seven-gene MLST gene detection for " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			get_st = Mlst(
				alleles = mlst_alleles,
				profiles = bcereus_path,
				perfect_matches = perfect_matches,
				final_results_directory = final_results_directory,
				prefix = prefix)


			mlst_final = get_st.at2st(mlst_alleles, bcereus_path, perfect_matches, final_results_directory, prefix)


	else:

		mlst_final = "(Seven-gene MLST not performed)"



	# perform panC phylogenetic group assignment using the adjusted, eight-group panC group assignment scheme
	if panC == "True":

		with importlib.resources.path("btyper3.seq_panC_db", "panC.fna") as panC_path:

			get_panC = Blast(
				task = "blastn",
				dbseqs = infile,
				fasta = panC_path,
				final_results_directory = final_results_directory,
				prefix = prefix,
				suffix = "panC",
				pthresh = 0,
				qthresh = 0,
				overlap = overlap,
				evalue = evalue)

			logging.info("Using blastn to identify panC in " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))

			panC_results = get_panC.run_blast("blastn", infile, panC_path, final_results_directory, prefix, "panC", evalue)
			panC_final = get_panC.parse_panC(panC_results)
			logging.info("Finished panC gene detection for " + prefix + " at " + now().strftime("%Y-%m-%d %H:%M"))


	else:

		panC_final = "(panC group assignment not performed)"



	# print results to a final results file
	get_final_results = FinalResults(
		final_results_directory = final_results_directory,
		infile = infile,
		prefix = prefix,
		species = final_species,
		subspecies = final_subspecies,
		geneflow = final_geneflow,
		typestrains = final_typestrains,
		anthracis = anthracis,
		emetic = emetic,
		nhe = nhe,
		hbl = hbl,
		cytK = cytK,
		sph = sph,
		cap = cap,
		has = has,
		bps = bps,
		bt_final = bt_final,
		mlst_final = mlst_final,
		panC_final = panC_final)

	get_final_results.print_final_results(final_results_directory, infile, prefix, final_species, final_subspecies, final_geneflow, final_typestrains, anthracis, emetic, nhe, hbl, cytK, sph, cap, has, bps, bt_final, mlst_final, panC_final)

	for blastdb_ext in ("nsq", "nin", "nhr", "ndb", "not", "ntf", "nto"):
		blastdb_file = "{}.{}".format(infile, blastdb_ext)
		if os.path.isfile(blastdb_file):
			os.remove(blastdb_file)

	logging.info("")
	logging.info("")
	logging.info("")
	logging.info("BTyper3 finished at " + now().strftime("%Y-%m-%d %H:%M"))
	logging.info("Report bugs/concerns to Laura M. Carroll, laura.carroll@embl.de\n")
	logging.info("Have a nice day!")



def main():

	# BTyper3 arguments

	parser = argparse.ArgumentParser(prog = "btyper3", usage = "btyper3 -i </path/to/genome.fasta> -o </path/to/output/directory/> [other options]")

	parser.add_argument("-i", "--input", help = "Path to input genome in fasta format", nargs = 1, required = True)

	parser.add_argument("-o", "--output", help = "Path to desired output directory", nargs = 1, required = True)

	parser.add_argument("--ani_species", help = "Optional argument; True or False; assign genome to a species using FastANI; default = True", nargs = "?", default = "True")

	parser.add_argument("--ani_subspecies", help = "Optional argument; True or False; assign genome to a subspecies, if relevant, using FastANI; default = True", nargs = "?", default = "True")

	parser.add_argument("--ani_geneflow", help = "Optional argument; True or False; assign genome to a pseudo-gene flow unit using the method described by Carroll, et al. using FastANI; default = False", nargs = "?", default = "False")

	parser.add_argument("--ani_typestrains", help = "Optional argument; True or False; calculate ANI values between the query genome relative to all B. cereus s.l. species type strain genomes using FastANI, and report the closest species type strain/highest ANI value; default = True", nargs = "?", default = "True")

	parser.add_argument("--virulence", help = "Optional argument; True or False; perform virulence gene detection (required if one wants to assign genomes to biovars Anthracis or Emeticus); default = True", nargs = "?", default = "True")

	parser.add_argument("--bt", help = "Optional argument; True or False; perform Bt toxin gene detection for cry, cyt, and vip genes (required if one wants to assign genomes to biovar Thuringiensis); default = True", nargs = "?", default = "True")

	parser.add_argument("--mlst", help = "Optional argument; True or False; assign genome to a sequence type (ST) using the seven-gene multi-locus sequence typing (MLST) scheme available in PubMLST; default = True", nargs = "?", default = "True")

	parser.add_argument("--panC", help = "Optional argument; True or False; assign genome to a phylogenetic group (Group I-VIII) using an adjusted, eight-group panC group assignment scheme; default = True", nargs = "?", default = "True")

	parser.add_argument("--virulence_db", help = "Optional argument for use with --virulence True; aa or nuc; database to use for virulence factor detection: aa for the amino acid sequence database, or nuc for the nucleotide sequence database; option aa uses translated nucleotide blast and allows for the detection of more remote homologs, but is slower than nuc, which uses blastn; default = aa", nargs = "?", default = "aa")

	parser.add_argument("--virulence_identity", help = "Optional argument for use with --virulence True; integer from 0 to 100; minimum percent amino acid/nucleotide identity threshold for a virulence gene to be considered present, depending on choice of --virulence_db aa or nuc, respectively; default = 70", nargs = "?", default = 70)

	parser.add_argument("--virulence_coverage", help = "Optional argument for use with --virulence True; integer from 0 to 100; minimum percent coverage threshold for a virulence gene to be considered present; default = 80", nargs = "?", default = 80)

	parser.add_argument("--bt_identity", help = "Optional argument for use with --bt True; integer from 0 to 100; minimum percent amino acid identity threshold for a Bt toxin gene to be considered present; default = 50", nargs = "?", default = 50)

	parser.add_argument("--bt_coverage", help = "Optional argument for use with --bt True; integer from 0 to 100; minimum percent coverage threshold for a Bt toxin gene to be considered present; default = 70", nargs = "?", default = 70)

	parser.add_argument("--bt_overlap", help = "Optional argument for use with --bt True; float from 0 to 1; specify maximum proportion of overlap for overlapping Bt toxin genes to be considered separate genes; Bt toxin genes below this threshold will be considered separate, while those above it will be considered overlapping, and only the top hit will be reported; default=0.7", nargs = "?", default = 0.7)

	parser.add_argument("--evalue", help = "Optional argument for use with --virulence True and/or --bt True; float >= 0; maximum blast e-value for a hit to be saved; note that if both --virulence True and --bt True, this e-value threshold will be applied to both analyses; default = 1e-5", nargs = "?", default = 1e-5)

	parser.add_argument("--download_mlst_latest", help = "Optional argument for use with --mlst True; True or False; download the latest version of the seven-gene multi-locus sequence typing (MLST) scheme available in PubMLST; if this is False, BTyper3 will search for the appropriate files in the seq_mlst_db directory; default = False", nargs = "?", default = "False")

	parser.add_argument("--version", action="version", version='%(prog)s {}'.format(__version__), help="Print version")

	args = parser.parse_args()

	run_pipeline(args)

if __name__ == "__main__":

	# run BTyper3

	main()
